lib/virt_autotest/esxi_utils.pm

```diff
@@ -0,0 +1,93 @@
+# SUSE's openQA tests
+#
+# Copyright © 2021 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+
+# Summary: Utilities for running ESXi commands
+# Maintainer: Nan Zhang <nan.zhang@suse.com>
+
+package virt_autotest::esxi_utils;
+
+use base Exporter;
+use Exporter;
+
+use base 'opensusebasetest';
+use strict;
+use warnings;
+use testapi;
+use utils;
+
+our @EXPORT = qw(
+  esxi_vm_get_vmid
+  esxi_vm_power_getstate
+  esxi_vm_power_off
+  esxi_vm_power_on
+  esxi_vm_power_reboot
+  esxi_vm_power_reset
+  esxi_vm_power_shutdown
+  esxi_vm_network_binding
+  esxi_vm_public_ip
+  get_host_timestamp
+  disable_vm_time_synchronization
+);
+
+sub esxi_vm_get_vmid {
+    my $vm_name = shift;
+    my $vim_cmd = "vim-cmd vmsvc/getallvms | grep -i $vm_name | awk -F' ' '{print \$1}'";
+    my (undef, $vmid) = console('svirt')->run_cmd($vim_cmd, domain => 'sshVMwareServer', wantarray => 1);
+    chomp($vmid);
+    return $vmid;
```

Comment:
It seems that console('svirt')->run_cmd is api which is called very often. So is there any possibility that it returns empty or not the thing you want ? If this happens, your test can run as expected.

Reply:
It's used to run server side commands, and it depends on the input parameters. I think it's unlikely to return an empty value if the input is correct.

Reply:
I see. Maybe this is just used this way for a long time. I still think it would be better to check whether the returned value is empty from run_cmd. This can be done in the function that tries to retrieve a value or the function that is going to use this value. Or in the run_cmd itself to check whether non-emtpy value is obtained. Or obtained empty value is allowed in run_cmd but it will fail or an option like the one used in script_output, proceed_on_failure, should b introduced. It gives me the feeling that run_cmd is not very good at error handling which should be enhanced from my POV.

Reply:
OK, I can understand your concern. It makes sense to check if any empty value which has impact during testing. Actually we already have part of inspection for the return value, such as esxi_vm_power_getstate() invoked in the function check_power_state(). The function run_cmd() is a wrapper for baseclass:run_ssh_cmd(),  seems we can't do more on it. So I would suggest to add related checks in the test function. As I said, the expected result we could get if the input is correct. My opinion is to ensure that the key input parameter is not empty string like $vm_id which is frequently used as a input parameter with run_cmd(), and such checking can be added at beginning of the tests.

Reply:
I see. You can go with your current way of checking. What I am concerned comes from what a API function is supposed to be.